File: backend/src/main/java/com/ocp/at/entity/ReceptionTravaux.java
package com.ocp.at.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Entité représentant la réception des travaux dans le workflow OCP.
 * Une seule réception est possible par Autorisation de Travail.
 */
@Entity
@Table(name = "receptions_travaux")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ReceptionTravaux {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "autorisation_travail_id", nullable = false, unique = true)
    private AutorisationTravail autorisationTravail;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "responsable_id")
    private Utilisateur responsable;

    private LocalDateTime dateReception;

    private LocalDateTime dateDebutTravauxReelle;

    private LocalDateTime dateFinTravauxReelle;

    @Column(columnDefinition = "TEXT")
    private String travauxRealises;

    @Column(nullable = false)
    @Builder.Default
    private Boolean travauxConformes = false;

    @Column(nullable = false)
    @Builder.Default
    private Boolean equipementRemisEnService = false;

    @Column(nullable = false)
    @Builder.Default
    private Boolean zoneNettoyee = false;

    @Column(nullable = false)
    @Builder.Default
    private Boolean consignationRetiree = false;

    @Column(nullable = false)
    @Builder.Default
    private Boolean essaisEffectues = false;

    @Column(columnDefinition = "TEXT")
    private String resultatEssais;

    @Column(columnDefinition = "TEXT")
    private String observations;

    @Column(columnDefinition = "TEXT")
    private String commentaireResponsable;

    // Signature manuscrite du responsable (stockée via StorageService)
    private String signaturePath;

    private LocalDateTime signatureDate;

    private String signatureBy;

    // Signature responsable pour clôture
    private String signatureResponsable;

    private LocalDateTime dateSignature;

    // Champs additionnels pour conformité
    private Boolean validee;

    private Boolean essaisConformes;

    private Boolean installationRemiseEnEtat;

    @Enumerated(EnumType.STRING)
    @Column(name = "resultat_reception", length = 30)
    @Builder.Default
    private com.ocp.at.entity.enums.ResultatReception resultatReception = com.ocp.at.entity.enums.ResultatReception.CONFORME;

    @Column(name = "reserves_description", columnDefinition = "TEXT")
    private String reservesDescription;

    @Column(name = "actions_correctives", columnDefinition = "TEXT")
    private String actionsCorrectives;

    @Column(name = "reception_conjointe_validee")
    @Builder.Default
    private Boolean receptionConjointeValidee = false;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "receptionTravaux", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<PhotoReception> photos = new ArrayList<>();

    @OneToMany(mappedBy = "receptionTravaux", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<HistoriqueReception> historiques = new ArrayList<>();
}

File: mobile/lib/features/at/data/models/reception_travaux.dart
/// Modèle ReceptionTravaux pour mobile.
/// Reflète ReceptionTravauxResponse et ReceptionTravauxRequest du backend.
library;

class ReceptionTravaux {
  final String id;
  final String autorisationTravailId;
  final DateTime? dateDebutTravauxReelle;
  final DateTime? dateFinTravauxReelle;
  final bool travauxConformes;
  final bool zoneNettoyee;
  final bool consignationRetiree;
  final bool equipementRemisEnService;
  final bool installationRemiseEnEtat;
  final bool essaisEffectues;
  final bool essaisConformes;
  final String? travauxRealises;
  final String? commentaireResponsable;
  final String? signatureResponsable;
  final DateTime? dateSignature;
  final bool cloturee;

  const ReceptionTravaux({
    required this.id,
    required this.autorisationTravailId,
    this.dateDebutTravauxReelle,
    this.dateFinTravauxReelle,
    this.travauxConformes = true,
    this.zoneNettoyee = true,
    this.consignationRetiree = true,
    this.equipementRemisEnService = true,
    this.installationRemiseEnEtat = true,
    this.essaisEffectues = true,
    this.essaisConformes = true,
    this.travauxRealises,
    this.commentaireResponsable,
    this.signatureResponsable,
    this.dateSignature,
    this.cloturee = false,
  });

  bool get isChecklistComplete =>
      travauxConformes &&
      zoneNettoyee &&
      consignationRetiree &&
      equipementRemisEnService &&
      installationRemiseEnEtat &&
      essaisEffectues &&
      essaisConformes;

  factory ReceptionTravaux.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic v) =>
        v == null ? null : DateTime.tryParse(v.toString());

    return ReceptionTravaux(
      id: json['id'] as String,
      autorisationTravailId: json['autorisationTravailId'] as String? ??
          (json['autorisationTravail'] is Map ? json['autorisationTravail']['id'] as String : ''),
      dateDebutTravauxReelle: parseDate(json['dateDebutTravauxReelle']),
      dateFinTravauxReelle: parseDate(json['dateFinTravauxReelle']),
      travauxConformes: json['travauxConformes'] as bool? ?? true,
      zoneNettoyee: json['zoneNettoyee'] as bool? ?? true,
      consignationRetiree: json['consignationRetiree'] as bool? ?? true,
      equipementRemisEnService: json['equipementRemisEnService'] as bool? ?? true,
      installationRemiseEnEtat: json['installationRemiseEnEtat'] as bool? ?? true,
      essaisEffectues: json['essaisEffectues'] as bool? ?? true,
      essaisConformes: json['essaisConformes'] as bool? ?? true,
      travauxRealises: json['travauxRealises'] as String?,
      commentaireResponsable: json['commentaireResponsable'] as String?,
      signatureResponsable: json['signatureResponsable'] as String?,
      dateSignature: parseDate(json['dateSignature']),
      cloturee: json['cloturee'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() => {
        'autorisationTravailId': autorisationTravailId,
        if (dateDebutTravauxReelle != null)
          'dateDebutTravauxReelle': _isoDate(dateDebutTravauxReelle!),
        if (dateFinTravauxReelle != null)
          'dateFinTravauxReelle': _isoDate(dateFinTravauxReelle!),
        'travauxConformes': travauxConformes,
        'zoneNettoyee': zoneNettoyee,
        'consignationRetiree': consignationRetiree,
        'equipementRemisEnService': equipementRemisEnService,
        'installationRemiseEnEtat': installationRemiseEnEtat,
        'essaisEffectues': essaisEffectues,
        'essaisConformes': essaisConformes,
        'travauxRealises': travauxRealises ?? '',
        'commentaireResponsable': commentaireResponsable ?? '',
      };

  static String _isoDate(DateTime d) =>
      '${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';
}
